test/playwright/visual.spec.js

```diff
@@ -0,0 +1,100 @@
+    context: async ({ }, use) => { // eslint-disable-line
+        const pathToExtension = path.join(root, 'ext');
+        const context = await chromium.launchPersistentContext('', {
+            // headless: false,
+            args: [
+                '--headless=new',
+                `--disable-extensions-except=${pathToExtension}`,
+                `--load-extension=${pathToExtension}`
+            ]
+        });
+        await use(context);
+        await context.close();
+    },
+    extensionId: async ({context}, use) => {
+        let [background] = context.serviceWorkers();
+        if (!background) {
+            background = await context.waitForEvent('serviceworker');
+        }
+
+        const extensionId = background.url().split('/')[2];
+        await use(extensionId);
+    }
+});
+const expect = test.expect;
+
+test('visual', async ({page, extensionId}) => {
+    // Install jmdict_english.zip
+    await page.goto(`chrome-extension://${extensionId}/settings.html`);
+    await page.waitForTimeout(500); // wait for welcome.html page to load, so we can then bringToFront() -- otherwise, the screenshot can hang
+    await expect(page.locator('id=dictionaries')).toBeVisible();
+    await page.bringToFront(); // improves performance
+    await expect.soft(page).toHaveScreenshot('settings-fresh.png', {maxDiffPixels: 100}); // maxDiffPixels is needed because the disk usage indicator is different per run
+    await page.waitForTimeout(200);
+    await page.locator('input[id="dictionary-import-file-input"]').setInputFiles(path.join(root, 'dictionaries/jmdict_english.zip'));
+    await expect(page.locator('id=dictionaries')).toHaveText('Dictionaries (1 installed, 1 enabled)', {timeout: 5 * 60 * 1000});
+    await expect.soft(page).toHaveScreenshot('settings-jmdict-loaded.png', {maxDiffPixels: 100}); // maxDiffPixels is needed because the disk usage indicator is different per run
+    await page.waitForTimeout(200);
+
+    const screenshot = async (doc_number, test_number, el, offset) => {
+        const test_name = 'doc' + doc_number + '-test' + test_number;
+
+        const box = await el.boundingBox();
+        if (box !== null) {
+            await page.mouse.move(box.x + offset.x, box.y + offset.y, {steps: 5}); // move on top of text
+        } else {
+            el.hover();
+        }
+
+        // note, it is not possible to use expect(...).toBeVisible() here because .yomichan-popop is in a closed shadow root
+        // plus, it takes time for the iframe in the closed shadow root to load
+        // in addition, we need to be super careful for there not to be a page navigation during the screenshot, otherwise it breaks
+        await page.waitForTimeout(500);
+
+        await page.bringToFront(); // required to ensure screenshot is taken quickly
+        await expect.soft(page).toHaveScreenshot(test_name + '.png');
+        await page.mouse.click(0, 0); // click away so popup disappears
+    };
+
+    // Load test-document1.html
```

Comment:
Both the procedures for taking screenshots for test-document1 and test-document2 are fairly similar. Could consider refactoring this a little by removing duplication

Reply:
Hmm, I already did by abstracting out the `screenshot` method -- other things like setting the viewport size and the specific selector are different per page. I think this is a decent balance personally, and would rather reevaluate the right abstraction if and when we add more pages.

Reply:
Fair enough. Could reevaluate once more pages get added